### ISSUE TITLE
[KNX] bad connection handling and resulting memory leak

### ISSUE THREAD

**Issue by muenning** [openhab/openhab2-addons#3134]
OpenHAB suffers a major memory leak caused by losing threads probably started by the KNX binding. 

## Expected Behavior
I expect OpenHAB to run reliable without the need to restart it several times a day.

## Current Behavior
After some running (12-24h with OpenHAB 1.8.3 and 4-12h with 2.x) the java thread count reaches its limit of about 8000, messages like `java.lang.OutOfMemoryError: unable to create new native thread` are filling the log and OpenHAB to basically stop working. After a restart everything is fine until the thread spawning starts again.

After some searching I think I have pinpointed the KNX binding to be the source of the problem. For unclear reasons the connection to the KNX router (MDT IP Router SCN-IP100.02 connected in TUNNEL mode as ROUTER mode is not working) is closed with a message like this logged:

```
2018-01-17 18:13:11.180 [WARN ] [.binding.knx.internal.bus.KNXBinding] - Value 'OFF' could not be sent to the KNX bus using datapoint 'command DP 10/3/56 WRNW, DPT main 0 id 1.001, low priority' - retrying one time: link closed, waiting for service ack
2018-01-17 18:13:11.180 [WARN ] [.binding.knx.internal.bus.KNXBinding] - Value 'OFF' could not be sent to the KNX bus using datapoint 'command DP 10/3/56 WRNW, DPT main 0 id 1.001, low priority' - retrying one time: link closed, connection closed
2018-01-17 18:13:11.179 [ERROR] [tuwien.auto.calimero                ] - calimero.link.192.168.0.16:3671: send error, closing link
```

After the first disconnect a continuous stream of connect attempts and failures are logged:

```
2018-01-17 18:16:10.066 [INFO ] [nx.internal.connection.KNXConnection] - Trying to (re-)connect to KNX...
2018-01-17 18:16:10.108 [WARN ] [nx.internal.connection.KNXConnection] - KNX link has been lost (reason: server request on object tunneling link link (closed) 192.168.0.16:3671 TP1 medium, device 0.0.0, hopcount 6)
2018-01-17 18:16:10.109 [INFO ] [nx.internal.connection.KNXConnection] - KNX link will be retried in 30 seconds
2018-01-17 18:16:10.113 [INFO ] [nx.internal.connection.KNXConnection] - Established connection to KNX bus on knx0:3671 in mode TUNNEL.
2018-01-17 18:16:10.114 [INFO ] [nx.internal.connection.KNXConnection] - Connected to KNX
2018-01-17 18:16:57.922 [INFO ] [nx.internal.connection.KNXConnection] - Trying to (re-)connect to KNX...
2018-01-17 18:16:58.110 [WARN ] [nx.internal.connection.KNXConnection] - KNX link has been lost (reason: server request on object tunneling link link (closed) 192.168.0.16:3671 TP1 medium, device 0.0.0, hopcount 6)
2018-01-17 18:16:58.113 [INFO ] [nx.internal.connection.KNXConnection] - KNX link will be retried in 30 seconds
2018-01-17 18:16:58.114 [ERROR] [.binding.knx.internal.bus.KNXBinding] - Received detach Event.
2018-01-17 18:16:59.526 [INFO ] [nx.internal.connection.KNXConnection] - Established connection to KNX bus on knx0:3671 in mode TUNNEL.
2018-01-17 18:16:59.529 [INFO ] [nx.internal.connection.KNXConnection] - Established connection to KNX bus on knx0:3671 in mode TUNNEL.
2018-01-17 18:16:59.530 [INFO ] [nx.internal.connection.KNXConnection] - Connected to KNX
2018-01-17 18:17:24.837 [WARN ] [nx.internal.connection.KNXConnection] - KNX link has been lost (reason: server request on object tunneling link link (closed) 192.168.0.16:3671 TP1 medium, device 0.0.0, hopcount 6)
2018-01-17 18:17:35.908 [INFO ] [nx.internal.connection.KNXConnection] - KNX link will be retried in 30 seconds
2018-01-17 18:17:28.115 [INFO ] [nx.internal.connection.KNXConnection] - Trying to (re-)connect to KNX...
```

This leads to one thread left over on every connection attempt. As the threads are not named I could verify this only by counting the log entries and comparing to the increase of java threads in the system. These connection attempts also use up all available channels on the KNX Router - ETS complains that there are no available connections left so I have to stop OpenHAB if I want to get a connection. Before this excessive re-connection attempts I can use ETS in parallel to running OpenHAB as expected.

I could not see any change in the behavior regardless of the `numberOfThreads` setting in knx.cfg. Basically no setting seems to have any effect on this.

By the way the disconnect message(s) are kinda bogus as KNX communication continues to work as expected even if OpenHAB pretends to be disconnected and tries to reconnect.

## Possible Solution
Check thread handling in KNX binding in regard of what happens when a connection is lost and retried. Terminating the left-over threads before reusing their handles should do the trick. Or some calimero clean-up is missed or the connection status is not queried correctly.
On the other hand re-sending the values to the bus seems to trigger the connection issue or at least makes it more probable. These resends happen regardless the `ignorelocalevents=true` setting.

## Steps to Reproduce (for Bugs)
Just let it run and watch the results. As this is probably related to the KNX hardware used (MDT IP Router SCN-IP100.02) and the amount of traffic (more than 100 KNX devices) it could be difficult to reproduce it on another or smaller system. On a smaller system it may just need to run longer to happen.

## Your Environment
openhab-1.8.3 with knx 1.9.0 binding
(all OpenHAB versions in between have the same problem)
openhab-2.2.0 with knx 1.11.0 binding
IP Router MDT SCN-IP100.02 connected in TUNNEL mode

my knx.cnf:
```
ip=192.168.0.16
ignorelocalevents=true
type=TUNNEL
localIp=192.168.0.1
autoReconnectPeriod=30
numberOfThreads=1
```

**Comment by kaikreuzer** [comment]
As this is about KNX 1.x binding, please report it at https://github.com/openhab/openhab1-addons/issues. Thanks!

**Comment by muenning** [comment]
Is there a KNX 2.x binding to try with?

**Comment by kaikreuzer** [comment]
Not yet: https://github.com/openhab/openhab2-addons/pull/2323
